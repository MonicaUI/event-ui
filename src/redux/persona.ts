import { getAttendeeId } from './selectors/currentRegistrant';
import { getIn } from 'icepick';
import * as InviteeStatus from 'event-widgets/utils/InviteeStatus';
import { loadRegistrationContent } from './actions';
import { APPROVAL_DENIED, PENDING_APPROVAL } from './website/registrationProcesses';
import { getRegistrationPathIdOrNull } from './selectors/currentRegistrationPath';

function getInviteeStatus(state) {
  return getIn(state, ['persona', 'inviteeStatus']);
}

export function isInviteePendingApproval(state: $TSFixMe): $TSFixMe {
  return InviteeStatus.PendingApproval === getInviteeStatus(state);
}

export function isInviteeApprovalDenied(state: $TSFixMe): $TSFixMe {
  return InviteeStatus.DeniedApproval === getInviteeStatus(state);
}

export function isInviteeCancelled(state: $TSFixMe): $TSFixMe {
  return InviteeStatus.Cancelled === getInviteeStatus(state);
}

const INVITEE_IDENTIFIED = 'event-guestside-site/persona/INVITEE_IDENTIFIED';
const UPDATE_INVITEE_STATUS_AND_CONTACT_ID = 'event-guestside-site/persona/UPDATE_INVITEE_STATUS_AND_CONTACT_ID';

export const createInviteeIdentifiedAction = (invitee: $TSFixMe): $TSFixMe => {
  return {
    type: INVITEE_IDENTIFIED,
    payload: invitee
  };
};

const createUpdateInviteeStatusAction = invitee => {
  return {
    type: UPDATE_INVITEE_STATUS_AND_CONTACT_ID,
    payload: invitee
  };
};

export const updateInviteeStatusAndContactIdInStore = (invitee: $TSFixMe) => {
  return async (dispatch: $TSFixMe): Promise<$TSFixMe> => {
    dispatch(createUpdateInviteeStatusAction(invitee));
  };
};

export const loadRegistrationContentForRegApproval = (inviteeStatusToLoad = null) => {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const inviteeStatus = inviteeStatusToLoad || getState().persona.inviteeStatus;
    if (inviteeStatus === InviteeStatus.PendingApproval) {
      await dispatch(loadRegistrationContent(PENDING_APPROVAL, getRegistrationPathIdOrNull(getState())));
    } else if (inviteeStatus === InviteeStatus.DeniedApproval) {
      await dispatch(loadRegistrationContent(APPROVAL_DENIED, getRegistrationPathIdOrNull(getState())));
    }
  };
};

export const identifyInvitee =
  (isContactIdRequired = false): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventPersonaClient },
      accessToken
    } = getState();
    const inviteeId = getState().userSession.inviteeId || getAttendeeId(getState());
    const invitee = await eventPersonaClient.identifyInvitee(accessToken, inviteeId, isContactIdRequired);
    if (invitee.inviteeStatus === InviteeStatus.PendingApproval) {
      await dispatch(loadRegistrationContent(PENDING_APPROVAL, getRegistrationPathIdOrNull(getState())));
    } else if (invitee.inviteeStatus === InviteeStatus.DeniedApproval) {
      await dispatch(loadRegistrationContent(APPROVAL_DENIED, getRegistrationPathIdOrNull(getState())));
    }
    dispatch(createInviteeIdentifiedAction(invitee));
  };

export const setPendingApprovalStatus =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    dispatch(createInviteeIdentifiedAction({ inviteeStatus: InviteeStatus.PendingApproval }));
    // eslint-disable-next-line @typescript-eslint/await-thenable
    await loadRegistrationContent(PENDING_APPROVAL, getRegistrationPathIdOrNull(getState()));
  };

export const personaReducer = (state = {}, action: $TSFixMe): $TSFixMe => {
  switch (action.type) {
    case INVITEE_IDENTIFIED:
      return {
        ...state,
        inviteeStatus: action.payload.inviteeStatus,
        contactId: action.payload.contactId
      };
    case UPDATE_INVITEE_STATUS_AND_CONTACT_ID:
      return {
        ...state,
        inviteeStatus: InviteeStatus.InviteeStatusById[action.payload.inviteeStatus],
        contactId: action.payload.contactId
      };
    default:
      return state;
  }
};
