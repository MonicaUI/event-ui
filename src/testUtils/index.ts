export const wait = (ms: $TSFixMe): $TSFixMe => new Promise((resolve): $TSFixMe => setTimeout(resolve, ms));

export function createRegistrationType(id = 'registrationTypeAId'): $TSFixMe {
  return {
    id,
    name: `name-${id}`,
    availableSessions: []
  };
}

export function createAdmissionItem(id = 'admissionItemAId', applicableContactTypes?: $TSFixMe): $TSFixMe {
  return {
    id,
    isOpenForRegistration: true,
    applicableContactTypes: applicableContactTypes || [],
    limitOptionalSessionsToSelect: false,
    availableOptionalSessions: [],
    associatedOptionalSessions: [],
    minimumNumberOfSessionsToSelect: 0,
    maximumNumberOfSessionsToSelect: undefined,
    type: 'AdmissionItem',
    code: `code-${id}`,
    name: `name-${id}`,
    description: `description-${id}`,
    fees: {},
    capacityId: `capacityId-${id}`
  };
}

export function createProductAdmissionItem(admissionItem: $TSFixMe): $TSFixMe {
  return {
    productId: admissionItem.id,
    productType: 'AdmissionItem',
    quantity: 1,
    requestedAction: 'REGISTER',
    registrationSourceType: 'Selected'
  };
}

export function createAutoAppliedAdmissionItem(admissionItem: $TSFixMe): $TSFixMe {
  return {
    productId: admissionItem.id,
    productType: 'AdmissionItem',
    quantity: 1,
    requestedAction: 'REGISTER'
  };
}

export function createSession(id = 'sessionItemAId'): $TSFixMe {
  return {
    id,
    isOpenForRegistration: true,
    isIncludedSession: false,
    startTime: '2000-03-20T13:00:00',
    endTime: '2000-03-20T14:00:00',
    locationName: 'Middle Earth',
    categoryId: 'id',
    capacity: 10,
    registeredCount: 2,
    associatedRegistrationTypes: [],
    type: 'Session',
    code: `code-${id}`,
    name: `name-${id}`,
    description: `description-${id}`,
    fees: {},
    capacityId: `capacityId-${id}`
  };
}

export function createQuantityItem(id = 'quantityItemAId'): $TSFixMe {
  return {
    id,
    associatedRegistrationTypes: [],
    isOpenForRegistration: true,
    capacityId: `capacityId-${id}`,
    status: 2,
    type: 'QuantityItem',
    code: `code-${id}`,
    name: `name-${id}`
  };
}

export function createQuantityItemAdvancedRules(id = 'quantityItemAdvancedRuleAId'): $TSFixMe {
  return {
    id,
    minQuantityAllowed: 1,
    quantityItem: [],
    maxQuantityAllowed: 3,
    name: `name-${id}`,
    warningMessage: `${id}-violated`,
    contactType: [],
    admissionItem: [],
    isActive: true
  };
}

export function createProductQuantityItem(quantityItem: $TSFixMe): $TSFixMe {
  return {
    productId: quantityItem.id,
    quantity: 2
  };
}

export function createDonationItem(id = 'donationItemAId'): $TSFixMe {
  return {
    id,
    associatedRegistrationTypes: [],
    isOpenForRegistration: true,
    status: 2,
    type: 'DonationItem',
    code: `code-${id}`,
    name: `name-${id}`
  };
}

export function createProductDonationItem(donationItem: $TSFixMe): $TSFixMe {
  return {
    productId: donationItem.id
  };
}

export function createSessionGroup(id = 'sessionGroupAId'): $TSFixMe {
  return {
    id,
    code: `code-${id}`,
    name: `name-${id}`,
    description: `description-${id}`,
    placementDateTime: '2000-03-20T13:00:00',
    categoryId: 'id',
    displayFormat: 1,
    displayOrder: 1,
    allowSessionDisplayFields: [1, 2, 8],
    isPlacementTimeDisplayed: true,
    isSessionSelectionRequired: true,
    isOpenForRegistration: true
  };
}

export function createProductSession(session: $TSFixMe): $TSFixMe {
  return {
    productId: session.id,
    productType: 'Session',
    quantity: 1,
    requestedAction: 'REGISTER',
    registrationSourceType: 'Selected'
  };
}

export function createCustomFieldAnswers(id = 'customFieldId'): $TSFixMe {
  return {
    questionId: id,
    answers: [
      {
        answerType: 'Choice',
        choice: `choice1-${id}`
      },
      {
        answerType: 'Choice',
        choice: `choice2-${id}`
      }
    ]
  };
}

export function createInvalidChildCustomFieldAnswers(id = 'childCustomFieldId', answers = []): $TSFixMe {
  return {
    questionId: id,
    answers
  };
}

export function createContactCustomFieldsMetadata(
  parentCustomFieldId = 'parentCustomFieldId',
  id = 'childCustomFieldId'
): $TSFixMe {
  return {
    contactCustomFields: [
      {
        question: {
          questionTypeInfo: {
            choices: [
              {
                id: 'childChoice1',
                text: `choice1-${id}`
              },
              {
                id: 'childChoice2',
                text: `choice2-${id}`
              },
              {
                id: 'childChoice3',
                text: `choice3-${id}`
              }
            ],
            linkLogic: {
              linkRules: [
                {
                  childChoices: [`choice3-${id}`]
                },
                {
                  childChoices: [`choice2-${id}`],
                  parentChoice: `choice1-${parentCustomFieldId}`
                },
                {
                  childChoices: [`choice1-${id}`],
                  parentChoice: `choice2-${parentCustomFieldId}`
                }
              ],
              parentQuestionLinkType: 'ContactCustomField',
              parentQuestionId: parentCustomFieldId
            }
          },
          id,
          text: 'child cust field with adv logic'
        }
      }
    ]
  };
}

export function createRegCartWithGuestRegistrations(): $TSFixMe {
  return {
    regCartId: 'd996d434-d088-44fb-8339-2831a2d0f93c',
    status: 'INPROGRESS',
    groupRegistration: false,
    eventRegistrations: {
      '00000000-0000-0000-0000-000000000001': {
        eventRegistrationId: '00000000-0000-0000-0000-000000000001',
        attendee: {
          isGroupMember: false
        },
        attendeeType: 'ATTENDEE',
        displaySequence: 1,
        productRegistrations: [
          {
            requestedAction: 'REGISTER',
            productType: 'AdmissionItem',
            productId: 'd4e7d4b3-77ac-4865-a30c-dac1f1055105',
            quantity: 1
          }
        ],
        requestedAction: 'REGISTER',
        primaryRegistrationId: '00000000-0000-0000-0000-000000000000',
        sessionRegistrations: {
          'fce781e0-b647-4548-af7e-badca1b0afad': {
            requestedAction: 'REGISTER',
            productId: 'fce781e0-b647-4548-af7e-badca1b0afad',
            registrationSourceType: 'Selected',
            includedInAgenda: false
          }
        },
        registrationTypeId: '8ddb813c-4fa1-4188-ac24-e54759b206bc',
        registrationPathId: 'fa7749e7-29db-44da-a6cc-bbe014599bb0'
      },
      '554ca936-5f54-4fd3-85be-376f27f4c4db': {
        eventRegistrationId: '554ca936-5f54-4fd3-85be-376f27f4c4db',
        attendee: {
          isGroupMember: false
        },
        attendeeType: 'GUEST',
        displaySequence: 2,
        productRegistrations: [
          {
            requestedAction: 'REGISTER',
            productType: 'AdmissionItem',
            productId: 'd4e7d4b3-77ac-4865-a30c-dac1f1055105',
            quantity: 1
          }
        ],
        requestedAction: 'UNREGISTER',
        primaryRegistrationId: '00000000-0000-0000-0000-000000000001',
        sessionRegistrations: {
          'fce781e0-b647-4548-af7e-badca1b0afad': {
            requestedAction: 'REGISTER',
            productId: 'fce781e0-b647-4548-af7e-badca1b0afad',
            registrationSourceType: 'Selected',
            includedInAgenda: false
          }
        },
        registrationTypeId: '8ddb813c-4fa1-4188-ac24-e54759b206bc',
        registrationPathId: 'fa7749e7-29db-44da-a6cc-bbe014599bb0'
      }
    }
  };
}

export function createConflictRegCartWithGuests(products?: $TSFixMe): $TSFixMe {
  const selectedAdmissionItem =
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    products && products.admissionItem
      ? createProductAdmissionItem(products.admissionItem)
      : createProductAdmissionItem(createAdmissionItem());
  const selectedSession =
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    products && products.session ? createProductSession(products.session) : createProductSession(createSession());
  return {
    regCartId: 'd996d434-d088-44fb-8339-2831a2d0f93c',
    status: 'INPROGRESS',
    groupRegistration: false,
    eventRegistrations: {
      eventRegistrationId: {
        eventRegistrationId: 'eventRegistrationId',
        attendee: {
          isGroupMember: false,
          airOptOutChoice: 'OPT_OUT'
        },
        attendeeType: 'ATTENDEE',
        displaySequence: 1,
        productRegistrations: [selectedAdmissionItem],
        requestedAction: 'REGISTER',
        primaryRegistrationId: '00000000-0000-0000-0000-000000000000',
        sessionRegistrations: {
          [selectedSession.productId]: selectedSession
        },
        registrationTypeId: '8ddb813c-4fa1-4188-ac24-e54759b206bc',
        registrationPathId: 'fa7749e7-29db-44da-a6cc-bbe014599bb0'
      },
      guestEventRegId: {
        eventRegistrationId: 'guestEventRegId',
        attendee: {
          isGroupMember: false,
          airOptOutChoice: 'OPT_OUT'
        },
        attendeeType: 'GUEST',
        displaySequence: 2,
        productRegistrations: [selectedAdmissionItem],
        requestedAction: 'REGISTER',
        primaryRegistrationId: 'eventRegistrationId',
        sessionRegistrations: {
          [selectedSession.productId]: selectedSession
        },
        registrationTypeId: 'registrationTypeAId',
        registrationPathId: 'fa7749e7-29db-44da-a6cc-bbe014599bb0'
      }
    }
  };
}

export function createConflictRegCartWithOptOutFlags(products?: $TSFixMe): $TSFixMe {
  const selectedAdmissionItem =
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    products && products.admissionItem
      ? createProductAdmissionItem(products.admissionItem)
      : createProductAdmissionItem(createAdmissionItem());
  const selectedSession =
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    products && products.session ? createProductSession(products.session) : createProductSession(createSession());
  return {
    regCartId: 'd996d434-d088-44fb-8339-2831a2d0f93c',
    status: 'INPROGRESS',
    groupRegistration: false,
    eventRegistrations: {
      eventRegistrationId: {
        eventRegistrationId: 'eventRegistrationId',
        attendee: {
          isGroupMember: false,
          airOptOutChoice: 'NOT_APPLICABLE'
        },
        attendeeType: 'ATTENDEE',
        displaySequence: 1,
        productRegistrations: [selectedAdmissionItem],
        requestedAction: 'REGISTER',
        primaryRegistrationId: '00000000-0000-0000-0000-000000000000',
        sessionRegistrations: {
          [selectedSession.productId]: selectedSession
        },
        registrationTypeId: '8ddb813c-4fa1-4188-ac24-e54759b206bc',
        registrationPathId: 'fa7749e7-29db-44da-a6cc-bbe014599bb0'
      },
      guestEventRegId: {
        eventRegistrationId: 'guestEventRegId',
        attendee: {
          isGroupMember: false,
          airOptOutChoice: 'NOT_APPLICABLE'
        },
        attendeeType: 'GUEST',
        displaySequence: 2,
        productRegistrations: [selectedAdmissionItem],
        requestedAction: 'REGISTER',
        primaryRegistrationId: 'eventRegistrationId',
        sessionRegistrations: {
          [selectedSession.productId]: selectedSession
        },
        registrationTypeId: 'registrationTypeAId',
        registrationPathId: 'fa7749e7-29db-44da-a6cc-bbe014599bb0'
      }
    }
  };
}

export function createConflictRegCartWithAirOptOutChoiceSetToBooked(products?: $TSFixMe): $TSFixMe {
  const selectedAdmissionItem =
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    products && products.admissionItem
      ? createProductAdmissionItem(products.admissionItem)
      : createProductAdmissionItem(createAdmissionItem());
  const selectedSession =
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    products && products.session ? createProductSession(products.session) : createProductSession(createSession());
  return {
    regCartId: 'd996d434-d088-44fb-8339-2831a2d0f93c',
    status: 'INPROGRESS',
    groupRegistration: false,
    eventRegistrations: {
      eventRegistrationId: {
        eventRegistrationId: 'eventRegistrationId',
        attendee: {
          isGroupMember: false,
          airOptOutChoice: 'BOOKED'
        },
        attendeeType: 'ATTENDEE',
        displaySequence: 1,
        productRegistrations: [selectedAdmissionItem],
        requestedAction: 'REGISTER',
        primaryRegistrationId: '00000000-0000-0000-0000-000000000000',
        sessionRegistrations: {
          [selectedSession.productId]: selectedSession
        },
        registrationTypeId: '8ddb813c-4fa1-4188-ac24-e54759b206bc',
        registrationPathId: 'fa7749e7-29db-44da-a6cc-bbe014599bb0'
      },
      guestEventRegId: {
        eventRegistrationId: 'guestEventRegId',
        attendee: {
          isGroupMember: false,
          airOptOutChoice: 'BOOKED'
        },
        attendeeType: 'GUEST',
        displaySequence: 2,
        productRegistrations: [selectedAdmissionItem],
        requestedAction: 'REGISTER',
        primaryRegistrationId: 'eventRegistrationId',
        sessionRegistrations: {
          [selectedSession.productId]: selectedSession
        },
        registrationTypeId: 'registrationTypeAId',
        registrationPathId: 'fa7749e7-29db-44da-a6cc-bbe014599bb0'
      }
    }
  };
}
