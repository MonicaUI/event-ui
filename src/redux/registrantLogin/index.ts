import status from './status';
import { combineReducers } from 'redux';

import form from './form';
import currentLogin from './currentLogin';

export default combineReducers({ form, status, currentLogin });
