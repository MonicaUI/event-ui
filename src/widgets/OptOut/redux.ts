/* eslint-disable import/no-cycle */

import { injectPageSnapshot } from '../../redux/actions';
import querystring from 'querystring';

/**
 * Opt-out and opt-in Redux module:
 * Event invitees can click "Opt Out" link in their emails to opt them out all communications from the event account
 */
const GET_OPT_OUT_STATUS = 'event-guestside-site/optOut/GET_OPT_OUT_STATUS';
const OPT_OUT_CONTACT = 'event-guestside-site/optOut/OPT_OUT_CONTACT';

const createGetOptedOutStatusAction = contactInfo => {
  return {
    type: GET_OPT_OUT_STATUS,
    payload: contactInfo
  };
};

export const getOptOutStatus =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventGuestClient },
      accessToken,
      event: { id: eventId }
    } = getState();
    let {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    if (contactId) {
      inviteeId = null;
    }
    const optedOutContact = await eventGuestClient.getOptOutStatus(accessToken, eventId, inviteeId, guestId, contactId);
    dispatch(createGetOptedOutStatusAction(optedOutContact));
  };

export const createOptedOutAction = (contactInfo: $TSFixMe): $TSFixMe => {
  return {
    type: OPT_OUT_CONTACT,
    payload: contactInfo
  };
};

export const optOutAction =
  (optOutStatus: $TSFixMe, fromConfirmPage: $TSFixMe): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventGuestClient },
      accessToken,
      event: { id: eventId }
    } = getState();
    let {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    if (contactId) {
      inviteeId = null;
    }
    // eslint-disable-next-line import/no-cycle
    const page = (await import(/* webpackChunkName optOut */ './content')).optInOutPageNew;
    if ((fromConfirmPage && optOutStatus) || !fromConfirmPage) {
      const optedOutContact = await eventGuestClient.optOut(
        accessToken,
        eventId,
        inviteeId,
        guestId,
        contactId,
        optOutStatus
      );
      dispatch(createOptedOutAction(optedOutContact));
    }
    dispatch(injectPageSnapshot(page));
  };

export const prepareForOptOutPageLoad =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    let page;
    if (!inviteeId && !guestId && !contactId) {
      page = (await import(/* webpackChunkName optOut */ './invalidInviteeContent')).invalidInviteeOptOut;
    } else {
      await dispatch(getOptOutStatus());
      if (getState() && getState().optOut && getState().optOut.optOutStatus) {
        // eslint-disable-next-line import/no-cycle
        page = (await import(/* webpackChunkName optOut */ './content')).optInOutPage;
      } else {
        // eslint-disable-next-line import/no-cycle
        page = (await import(/* webpackChunkName optOut */ './content')).confirmationPage;
      }
    }
    dispatch(injectPageSnapshot(page));
  };

export const optOutReducer = (state = {}, action: $TSFixMe): $TSFixMe => {
  switch (action.type) {
    case GET_OPT_OUT_STATUS:
    case OPT_OUT_CONTACT:
      return {
        ...state,
        ...action.payload
      };
    default:
      return state;
  }
};
