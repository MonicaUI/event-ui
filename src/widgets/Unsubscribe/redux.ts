import { injectPageSnapshot } from '../../redux/actions';
import querystring from 'querystring';

/**
 * Unsubscribe and re-subscribe Redux module:
 * Event invitees can click "Unsubscribe" link in their emails to opt them out from the event emails communication
 */
const GET_UNSUBSCRIBE_STATUS = 'event-guestside-site/unsubscribe/GET_UNSUBSCRIBE_STATUS';
const SET_UNSUBSCRIBE_STATUS = 'event-guestside-site/unsubscribe/SET_UNSUBSCRIBE_STATUS';

const createGetUnsubscribeStatusAction = (unsubscribedContact, optOutStatus) => {
  return {
    type: GET_UNSUBSCRIBE_STATUS,
    payload: {
      ...optOutStatus,
      unsubscribeStatus: !unsubscribedContact.subscribed
    }
  };
};

export const getUnsubscribeStatus =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventEmailClient, eventGuestClient },
      accessToken,
      event: { id: eventId },
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const encodedGuestId = queryParams.g;
    const unsubscribedContact = await eventEmailClient.getSubscriptionStatus(eventId, inviteeId, encodedGuestId);
    // optOutStatus object has invitee/guest and planner infomation for display
    const optOutStatus = await eventGuestClient.getOptOutStatus(accessToken, eventId, inviteeId, encodedGuestId);
    dispatch(createGetUnsubscribeStatusAction(unsubscribedContact, optOutStatus));
  };

const createUnsubscribeAction = unsubscribeStatus => {
  return {
    type: SET_UNSUBSCRIBE_STATUS,
    payload: { unsubscribeStatus }
  };
};

export const unsubscribeAction =
  (unsubscribeStatus: $TSFixMe): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventEmailClient },
      event: { id: eventId },
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const encodedGuestId = queryParams.g;
    const subscribed = !unsubscribeStatus;
    await eventEmailClient.setSubscriptionStatus(eventId, inviteeId, encodedGuestId, subscribed);
    dispatch(createUnsubscribeAction(unsubscribeStatus));
  };

export const prepareForUnsubscribePageLoad =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const encodedGuestId = queryParams.g;
    let unsubscribePage;
    if (!inviteeId && !encodedGuestId) {
      unsubscribePage = (await import(/* webpackChunkName: "unsubscribe" */ './invalidInviteeContent'))
        .invalidInviteeUnsubscribePage;
    } else {
      await dispatch(getUnsubscribeStatus());
      // eslint-disable-next-line import/no-cycle
      unsubscribePage = (await import(/* webpackChunkName: "unsubscribe" */ './content')).unsubscribePage;
    }
    dispatch(injectPageSnapshot(unsubscribePage));
  };

export const unsubscribeReducer = (state = {}, action: $TSFixMe): $TSFixMe => {
  switch (action.type) {
    case GET_UNSUBSCRIBE_STATUS:
    case SET_UNSUBSCRIBE_STATUS:
      return {
        ...state,
        ...action.payload
      };
    default:
      return state;
  }
};
