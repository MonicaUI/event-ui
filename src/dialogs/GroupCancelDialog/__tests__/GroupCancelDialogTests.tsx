import React from 'react';
import dialogContainer, * as dialogContainerActions from 'nucleus-guestside-site/src/redux/modules/dialogContainer';
import DialogContainer from 'nucleus-guestside-site/src/containers/DialogContainer';
import createStoreWithMiddleware from 'nucleus-guestside-site/src/redux/devCreateStoreWithMiddleware';
import { combineReducers } from 'redux';
import { openGroupCancelRegistrationDialog } from '..';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import EventSnapshot from '../../../../fixtures/EventSnapshot.json';
import { wait } from '../../../testUtils';
import registrantLogin from '../../../redux/registrantLogin';
import * as InviteeStatus from 'event-widgets/utils/InviteeStatus';
import { personaReducer } from '../../../redux/persona';

dialogContainerActions.showLoadingDialog = jest.fn(() => () => {});
dialogContainerActions.hideLoadingDialog = jest.fn(() => () => {});
dialogContainerActions.hideLoadingOnError = jest.fn(() => () => {});
const eventGuestClient = {};
const identifiedInvitee = { inviteeStatus: InviteeStatus.NoResponse };
const eventPersonaClient = {
  identifyInvitee: jest.fn().mockReturnValue(
    new Promise(resolve => {
      resolve(identifiedInvitee);
    })
  )
};
const store = createStoreWithMiddleware(
  combineReducers({
    dialogContainer,
    text: (x = {}) => x,
    website: (x = {}) => x,
    pathInfo: (x = {}) => x,
    clients: (x = {}) => x,
    event: (x = {}) => x,
    userSession: (x = {}) => x,
    defaultUserSession: (x = {}) => x,
    plannerRegSettings: (x = {}) => x,
    accessToken: (x = {}) => x,
    registrantLogin,
    registrationForm: (x = {}) => x,
    appData: (x = {}) => x,
    persona: (x = {}) => x
  }),
  {
    text: {
      translate: (resx, tokens) => (tokens ? `${resx}:${JSON.stringify(tokens)}` : resx)
    },
    website: EventSnapshot.eventSnapshot.siteEditor.website,
    clients: { eventGuestClient, eventPersonaClient },
    event: {
      id: 'eventId'
    },
    defaultUserSession: {
      isPreview: false,
      isPlanner: false
    },
    plannerRegSettings: {
      exitURL: 'url'
    },
    accessToken: 'accessToken',
    registrantLogin: {
      form: {
        firstName: 'firstName',
        lastName: 'lastName',
        emailAddress: 'emailAddress',
        confirmationNumber: 'confirmationNumber'
      },
      status: {
        login: {},
        resendConfirmation: {}
      }
    },
    appData: {
      registrationSettings: {
        registrationPaths: {
          todoThisShouldBeRegPathId: {
            id: 'todoThisShouldBeRegPathId',
            isDefault: true
          }
        }
      }
    },
    persona: personaReducer
  }
);

describe('GroupCancelDialog', () => {
  const dialog = mount(
    <Provider store={store}>
      <DialogContainer spinnerMessage="spinnerMessage" message="message" />
    </Provider>
  );

  test('matches snapshot when opened', () => {
    store.dispatch(
      openGroupCancelRegistrationDialog(() => () => true, {
        title: 'Group Cancel',
        style: EventSnapshot.eventSnapshot.siteEditor.website.theme.global
      })
    );
    expect(dialog).toMatchSnapshot();
  });
  test('closes group cancel dialog', () => {
    dialog.update();
    dialog.find('[data-cvent-id="no-submit-button"]').hostNodes().simulate('click');
    expect(dialog).toMatchSnapshot();
  });

  test('Exits registration normally on clicking yes', async () => {
    store.dispatch(
      openGroupCancelRegistrationDialog(() => () => true, {
        title: 'Group Cancel',
        style: EventSnapshot.eventSnapshot.siteEditor.website.theme.global
      })
    );
    dialog.update();
    dialog.find('[data-cvent-id="yes-submit-button"]').hostNodes().simulate('click');
    await wait(1000);
  });
});
