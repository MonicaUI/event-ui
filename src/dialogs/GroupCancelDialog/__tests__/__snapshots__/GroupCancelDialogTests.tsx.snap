// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupCancelDialog closes group cancel dialog 1`] = `
Array [
  <div
    class=""
  >
    <div
      class=""
    />
    <span
      class="mockNucleusCoreTransition"
    />
  </div>,
]
`;

exports[`GroupCancelDialog matches snapshot when opened 1`] = `
Array [
  <div
    class=""
  >
    <div
      class=""
    />
    <span
      class="mockNucleusCoreTransition"
    >
      <div
        class=""
      >
        <div
          aria-label="Dialog"
          role="dialog"
          style="outline: none;"
          tabindex="0"
        >
          <div
            class=""
            style="background-color: rgb(243, 59, 21); border-radius: 100px; position: absolute; left: 512px; top: 384px;"
          >
            <div
              class=""
            />
            <div
              class=""
            >
              <div
                class="dialogContainer"
              >
                <div
                  class="dialogChildren"
                >
                  <div
                    data-cvent-id="group-cancel-registration"
                  >
                    <div>
                      <div
                        aria-describedby="dialogHeader"
                        class=""
                        style="background-color: rgb(221, 226, 230); border-radius: 0;"
                      >
                        <div
                          class=""
                          id="dialogHeader"
                          style="background-color: rgb(249, 32, 242); padding: 15px 1.5625rem 15px 1.5625rem;"
                        >
                          <h2
                            class=""
                            style="color: rgb(255, 255, 255); font-family: Comic Sans MS, Comic Sans, Chalkboard, Marker Felt, sans-serif; font-size: 3.5625rem; font-style: normal; font-weight: 700; line-height: 1.3;"
                          >
                            <div
                              class=""
                              data-cvent-id="close"
                              role="button"
                              tabindex="0"
                            >
                              <span
                                class="container"
                              >
                                <span
                                  aria-label="EventGuestSide_StandardDialog_FallbackText_Close__resx"
                                  class=""
                                  role="img"
                                />
                              </span>
                            </div>
                            EventGuestsideSite_GroupCancelDialog_Title__resx
                          </h2>
                        </div>
                      </div>
                      <div
                        class=""
                        name="messageContainer"
                        style="background-color: rgb(243, 59, 21); padding: 0px 1.25rem 1.875rem 1.25rem;"
                        tabindex="-1"
                      >
                        <div
                          class=""
                          style="background-color: rgb(243, 59, 21); padding: 0px 1.25rem 1.875rem 1.25rem;"
                        >
                          <div
                            data-cvent-id="confirmation-form"
                          >
                            <div
                              class="cancelRegistrationModalText"
                              style="font-family: PT Sans, sans-serif; font-weight: 400; font-size: 16px; color: rgb(39, 63, 105); line-height: 1.3; font-style: normal; background-color: transparent; border-radius: 0; padding: 0px 0px 0px 0px;"
                            >
                              EventGuestsideSite_GroupCancelDialog_InstructionalText__resx
                            </div>
                            <button
                              aria-label="EventWidgets_GenericText_No__resx"
                              class=""
                              data-cvent-id="no-submit-button"
                              title="EventWidgets_GenericText_No__resx"
                              type="button"
                            >
                              <span>
                                EventWidgets_GenericText_No__resx
                              </span>
                            </button>
                            <button
                              aria-label="EventWidgets_GenericText_Yes__resx"
                              class=""
                              data-cvent-id="yes-submit-button"
                              title="EventWidgets_GenericText_Yes__resx"
                              type="button"
                            >
                              <span>
                                EventWidgets_GenericText_Yes__resx
                              </span>
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </span>
  </div>,
]
`;
